// dtslint/bin/rules/noRedundantUndefinedRule.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rule = void 0;
const Lint = __importStar(require("tslint"));
const ts = __importStar(require("typescript"));
const util_1 = require("../util");
class Rule extends Lint.Rules.AbstractRule {
    apply(sourceFile) {
        return this.applyWithFunction(sourceFile, walk);
    }
}
exports.Rule = Rule;
Rule.metadata = {
    ruleName: "no-redundant-undefined",
    description: "Forbids optional parameters/properties to include an explicit `undefined` in their type.",
    optionsDescription: "Not configurable.",
    options: null,
    type: "style",
    typescriptOnly: true,
};
function failureString(container) {
    const desc = container.kind === ts.SyntaxKind.Parameter ? "Parameter" : "Property";
    return util_1.failure(Rule.metadata.ruleName, `${desc} is optional, so no need to specify \`undefined\` as a possible value.`);
}
function walk(ctx) {
    ctx.sourceFile.forEachChild(function recur(node) {
        if (node.kind === ts.SyntaxKind.UndefinedKeyword
            && ts.isUnionTypeNode(node.parent)
            && isOptionalParent(node.parent.parent)) {
            ctx.addFailureAtNode(node, failureString(node.parent.parent));
        }
        node.forEachChild(recur);
    });
}
function isOptionalParent(node) {
    switch (node.kind) {
        case ts.SyntaxKind.Parameter:
        case ts.SyntaxKind.PropertyDeclaration:
        case ts.SyntaxKind.PropertySignature:
            return node.questionToken !== undefined;
        default:
            return false;
    }
}
